File: node-ts/api-exemplo/src/tests/user.test.ts
import { describe, it, expect } from 'vitest';
import { findById, findAll, create } from '../models/user.model.js';

describe('user model', () => {
  it('lista usuários iniciais', () => {
    expect(findAll().length).toBe(2);
  });

  it('acha usuário existente por id', () => {
    expect(findById(1)?.name).toBe('Ada Lovelace');
  });

  it('retorna undefined pra id inexistente', () => {
    expect(findById(999)).toBeUndefined();
  });

  it('cria um novo usuário', () => {
    const user = create({ name: 'Grace Hopper', email: 'grace@example.com' });
    expect(user.id).toBeGreaterThan(0);
    expect(findById(user.id)?.name).toBe('Grace Hopper');
  });
});

File: golang/api-exemplo/models/user_test.go
package models

import "testing"

func TestFindAll(t *testing.T) {
	if len(FindAll()) != 2 {
		t.Errorf("esperava 2 usuários iniciais, veio %d", len(FindAll()))
	}
}

func TestFindByIDExistente(t *testing.T) {
	user, ok := FindByID("1")
	if !ok || user.Name != "Ada Lovelace" {
		t.Errorf("esperava Ada Lovelace, veio %+v (ok=%v)", user, ok)
	}
}

func TestFindByIDInexistente(t *testing.T) {
	_, ok := FindByID("999")
	if ok {
		t.Error("esperava não encontrar o usuário 999")
	}
}

func TestCreate(t *testing.T) {
	user, err := Create("Grace Hopper", "grace@example.com")
	if err != nil {
		t.Fatalf("não esperava erro, veio: %v", err)
	}
	if user.ID == 0 {
		t.Error("esperava um ID gerado")
	}
}

func TestCreateSemNome(t *testing.T) {
	_, err := Create("", "grace@example.com")
	if err == nil {
		t.Error("esperava erro por falta de name")
	}
}
